core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -271,12 +271,27 @@ CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>> p
                         // Maybe check if no records are acquired and we want to retry replica
                         // manager fetch. Depends on the share partition manager implementation,
                         // if we want parallel requests for the same share partition or not.
-                        partitionData
-                            .setPartitionIndex(topicIdPartition.partition())
-                            .setRecords(fetchPartitionData.records)
-                            .setErrorCode(fetchPartitionData.error.code())
-                            .setAcquiredRecords(acquiredRecords)
-                            .setAcknowledgeErrorCode(Errors.NONE.code());
+                        if (fetchPartitionData.error.code() == Errors.OFFSET_OUT_OF_RANGE.code()) {
+                            // In case we get OFFSET_OUT_OF_RANGE error, that's because the LSO is later than the fetch offset.
+                            // So, we would update the start and end offset of the share partition and still return an empty
+                            // response and let the client retry the fetch. This way we do not lose out on the data that
+                            // would be returned for other share partitions in the fetch request.
+                            sharePartition.updateOffsetsOnLsoMovement();
+                            partitionData
+                                .setPartitionIndex(topicIdPartition.partition())
+                                .setRecords(fetchPartitionData.records)
```

Comment:
Will there be any records in case of exception? If not then explicitly fetching records from `fetchPartitionData` seems incorrect rather explicitly set empty collection.

Reply:
There won't be any records in such case, we should populate records to `null` in such case.